### PR TITLE
COSBETA-87: Populate formulation information from notification file import

### PR DESCRIPTION
**Pull request by aidayounesi**
Import the following data into a new notification from an imported file:
    - Category
    - Notification type (predefined, ranges, exact)
    - Formulation information (i.e. list of ranges/exact values, or flag saying data is in an attachment)
    - Trigger rule answers

### CODE REVIEW

File: cosmetics-web/app/analyzers/concerns/cpnp_static_files.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+require 'zlib'
+
+module CpnpStaticFiles
+  def is_static_file(file)
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/concerns/cpnp_static_files.rb]: Sure, thanks

File: cosmetics-web/app/analyzers/concerns/cpnp_static_files.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    Rails.application.config.cpnp_static_file_names.include? get_file_name(file)
+  end
+
+  def static_file_contents_differs(file)
+    file_hash = Zlib::crc32(file.get_input_stream.read)
+    expected_hash = Rails.application.config.cpnp_static_file_hashes[get_file_name(file)]
+    file_hash != expected_hash
+  end
+
+  def get_file_name(file)
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/concerns/cpnp_static_files.rb]: Sure, thanks

File: cosmetics-web/app/analyzers/read_data_analyzer.rb [hunk 3]
```diff
@@ -42,12 +44,18 @@ def create_notification_from_file
       @notification_file.update(upload_error: :product_file_not_found)
     rescue UnexpectedFileError
       @notification_file.update(upload_error: :unzipped_files_not_xml)
+    rescue UnexpectedStaticFilesError
+      @notification_file.update(upload_error: :static_files_differs)
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/read_data_analyzer.rb]: Sure, thanks

File: cosmetics-web/app/analyzers/read_data_analyzer.rb [hunk 4]
```diff
@@ -62,22 +70,36 @@ def get_product_xml_file
...
-    /[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{16}.*\.xml/
+  def file_is_product_xml?(file)
+    file.name&.match?(get_xml_file_name_regex)
   end
 
   def file_is_pdf?(file)
     file.name&.match?(/.*\.pdf/)
   end
 
+  def get_xml_file_name_regex
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/read_data_analyzer.rb]: Sure, thanks

File: cosmetics-web/app/helpers/cpnp_helper.rb [hunk 5]
```diff
@@ -0,0 +1,347 @@
+module CpnpHelper
+  def get_category_name(category)
+    @category_name[category]
+  end
+
+  def get_frame_formulation_name(frame_formulation)
+    @frame_formulation_name[frame_formulation]
+  end
+
+  @category_name = {
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/helpers/cpnp_helper.rb]: Changed them to constants. freeze is added by robocup auto correction and based on ruby documentations it prevent them from being modified.

File: cosmetics-web/app/models/component.rb [hunk 6]
```diff
@@ -1,8 +1,14 @@
 class Component < ApplicationRecord
   include AASM
+  include UkNotificationInfo
 
   belongs_to :notification
 
+  has_many :exact_formulas, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/component.rb]: Yes, it is, Dave provided an example zip file that a component has:
```
											<sanco-xmlgate:exactFormulaList>
												
												<sanco-xmlgate:formula>
													<sanco-xmlgate:inciName>B</sanco-xmlgate:inciName>
													<sanco-xmlgate:quantity>99</sanco-xmlgate:quantity>										
												</sanco-xmlgate:formula>
												
												<sanco-xmlgate:formula>
													<sanco-xmlgate:inciName>A</sanco-xmlgate:inciName>
													<sanco-xmlgate:quantity>1</sanco-xmlgate:quantity>										
												</sanco-xmlgate:formula>
																
											</sanco-xmlgate:exactFormulaList>
```

File: cosmetics-web/app/models/concerns/cpnp_notification_info.rb [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  def get_category(id)
+    CATEGORY_ID[id]
+  end
+
+  def get_frame_formulation(id)
+    FRAME_FORMULATION_ID[id]
+  end
+
+  NOTIFICATION_TYPE_ID = {
+      1 => :predefined,
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/concerns/cpnp_notification_info.rb]: At least RubyMine is not happy about it ('Error: Expected , ')

[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/concerns/cpnp_notification_info.rb]: https://stackoverflow.com/questions/37465921/ruby-integer-as-key-in-hash

File: cosmetics-web/config/initializers/cpnp_static_files.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+require 'zlib'
+
+Rails.application.config.cpnp_static_file_names = %w(categories frameFormulation).freeze
+
+Rails.application.config.cpnp_static_file_hashes = Hash[Rails.application.config.cpnp_static_file_names.map do |filename|
+  file_content = File.read(Rails.root.join('app', 'assets', 'files', 'cpnp', filename + ".xml"))
+  [filename, Zlib::crc32(file_content)]
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/config/initializers/cpnp_static_files.rb]: Sure, thanks

File: cosmetics-web/app/models/cpnp_export.rb [hunk 9]
```diff
@@ -16,19 +17,78 @@ def cpnp_reference
...
   end
 
   def shades
-    @xml_doc.xpath('//currentVersion/generalInfo/productNameList/productName/shade').first&.text
+    current_version_info_node.xpath('.//shade').first&.text
+  end
+
+  def components
+    current_version_component_lists_node.xpath('.//component').collect do |component_node|
+      Component.create(notification_type: notification_type(component_node),
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/cpnp_export.rb]: Component's name and shades are captured now. 
I needed to remove validation of shades on component model. Please refer to https://regulatorydelivery.atlassian.net/browse/COSBETA-108 to see full TODO lists about shades.

File: cosmetics-web/config/initializers/cpnp_static_files.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+require 'zlib'
+
+Rails.application.config.cpnp_static_file_names = %w(categories frameFormulation).freeze
+
+Rails.application.config.cpnp_static_file_hashes = Hash[Rails.application.config.cpnp_static_file_names.map do |filename|
+  file_content = File.read(Rails.root.join('app', 'assets', 'files', 'cpnp', filename + ".xml"))
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/config/initializers/cpnp_static_files.rb]: Sure, thanks

File: cosmetics-web/app/models/concerns/uk_notification_info.rb [hunk 11]
```diff
@@ -0,0 +1,458 @@
+# rubocop:disable Metrics/BlockLength
+module UkNotificationInfo
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/concerns/uk_notification_info.rb]: That's a good idea. NotificationTypes, NotificationCategories, NotificationFrameFormulations and NotificationTriggerRules modules are created. And all of them are included in a module called NotificationProperties which is included in Component model itself.

File: cosmetics-web/app/models/concerns/cpnp_notification_info.rb [hunk 12]
```diff
@@ -0,0 +1,357 @@
+module CpnpNotificationInfo
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/concerns/cpnp_notification_info.rb]: Similar to previous comment: CpnpNotificationTypes, CpnpNotificationCategories, CpnpNotificationFrameFormulations and CpnpNotificationTriggerRules modules are created. And all of them are included in a module called CpnpNotificationProperties which is included in CpnpExport model itself.

File: cosmetics-web/lib/tasks/cpnp_import.rake [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    print_three_mapping_data_structures('category', 'categories', 'category')
+  end
+
+  task :categories_parent do
+    parent_hash_command = start_hash_command("PARENT_OF_CATEGORY")
+
+    xml_file_content = File.read(Rails.root.join('app', 'assets', 'files', 'cpnp', 'categories.xml'))
+    xml_doc = Nokogiri::XML(xml_file_content)
+
+    underlined_names_set = Set.new
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/lib/tasks/cpnp_import.rake]: Sure, thanks

File: cosmetics-web/app/analyzers/read_data_analyzer.rb [hunk 14]
```diff
@@ -62,22 +70,36 @@ def get_product_xml_file
     end
   end
 
-  def file_is_product_xml?(file)
-    file.name&.match?(get_xml_file_name_regex)
+  def invalid_static_files(files)
+    files.any? do |file|
+      differs = is_static_file(file) && static_file_contents_differs(file)
+      if differs
+        Rails.logger.error "***** WARNING - different static file was detected! *****"
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/read_data_analyzer.rb]: Sure, thanks

File: cosmetics-web/config/locales/en.yml [hunk 15]
```diff
@@ -29,6 +29,7 @@ en:
         postal_code: "Postcode"
       notification_file:
         upload_errors:
+          static_files_differs: "Unexpected files in the ZIP file"
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/config/locales/en.yml]: Sure, thanks

File: cosmetics-web/app/models/range_formula.rb [hunk 16]
```diff
@@ -0,0 +1,3 @@
+class RangeFormula < ApplicationRecord
+  belongs_to :component
```
[the remark this answers is not in the document]
Comment (aidayounesi) [reply cosmetics-web/app/models/range_formula.rb]: Yes we need an enum for it too. There's a static file called units_xxxx.xml that I can parse.

File: cosmetics-web/app/analyzers/read_data_analyzer.rb [hunk 17]
```diff
@@ -67,12 +69,18 @@ def create_notification_from_file
...
+    rescue UnknownError
+      @notification_file.update(upload_error: :unknown_error)
     end
   end
 
   def get_product_xml_file
     download_blob_to_tempfile do |zip_file|
       Zip::File.open(zip_file.path) do |files|
+        if invalid_static_files(files)
+          raise UnknownError, "UnknownError - a different static file was detected!"
```
Comment (tom-bell-softwire) [inline cosmetics-web/app/analyzers/read_data_analyzer.rb]: Sorry, my last comment wasn't very clear! I think it's better to raise a specific `UnexpectedStaticFilesError` error here, but when we `rescue` it, the `upload_error` value should be `:unknown_error` (which is what you now have above).

File: cosmetics-web/app/models/component.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (tom-bell-softwire) [reply cosmetics-web/app/models/component.rb]: That's weird! Oh well...

File: cosmetics-web/app/models/concerns/cpnp_notification_info.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (tom-bell-softwire) [reply cosmetics-web/app/models/concerns/cpnp_notification_info.rb]: Sure – thanks.

File: cosmetics-web/app/analyzers/read_data_analyzer.rb (hunk 17 above)
Comment (aidayounesi) [reply cosmetics-web/app/analyzers/read_data_analyzer.rb]: Sure thanks

**Pull request merged**